src/server/activeobjectmgr.cpp

```diff
@@ -118,6 +118,16 @@ void ActiveObjectMgr::removeObject(u16 id)
 	}
 }
 
+void ActiveObjectMgr::invalidateActiveObjectObserverCaches()
+{
+	for (auto &activeObject : m_active_objects.iter()) {
+		ServerActiveObject *obj = activeObject.second.get();
+		if (!obj)
+			continue;
+		obj->invalidateFinalObservers();
```

Comment:
> would it work to reset this when the object is being stepped anyway?

It would work, yes, but it is likely to lead to unnecessary recalculations. Resetting these upfront guarantees time linear in the number of objects and observers.

> alternatively maybe parents could recursively propagate changes down to all their children

I considered this, but that would negatively affect the time complexity of methods like `set_attach`. I'm not sure whether that would matter, but by recalculating this during steps, the asymptotic time complexity of steps remains the same, since those are already O(nm) due to having to iterate over every entity for every player.

I don't think it makes a big difference either way.

Reply:
n = number of objects, m = number of parents (or children?) per object

current PR:
* extra `O(n)` loop every step, but it's trivial
* `O(nm)` loop every step¹ to recalculate

what I propose:
* no loop to reset
* `O(m)` only when observability changed, an object is attached, detached or removed

surely this makes a difference?

¹: due to `isEffectivelyObservedBy` in `ServerEnvironment::getRemovedActiveObjects`

Reply:
n = number of objects, m = number of players (more or less an upper bound on the number of observers, if the feature is used reasonably, though we could also enforce observer sets only containing online players). The current server step code is O(nm) at least, because for each player it has to go through all objects. Hence we don't worsen the situation asymptotically currently.

With your proposal (propagating updates to children), the constant factor of server steps would be lower. But instead changing observability, detaching or attaching objects ("updates") get more expensive. This is worse than O(k) where k is the number of descendants of an object: Note that for each descendant you also have to compute the O(m) set intersection with its parent. So you get O(mk) for a single "update". If each object updates in a step, and each object has, say sqrt(n) children on average (worst case), this gives us O(n sqrt(n) m) for a server step.

You could also recompute the entire parent chain whenever you update a child. This might be better for flat hierarchies or if updates tend to affect parents rather than children. But asymptotically it is no better in the worst case - it's O(n sqrt(n) m) as well.

(There is the advantage that `get_effective_observers` would be O(m), but I consider that a small one.)

---

In practice, I think none of this is likely to matter. Attachment hierarchies are very flat (usually probably only 1, maybe 2 parents, tops maybe 10 parents); many (most?) objects are not attached at all. I would expect observers to not be used excessively, so there will only be a few objects where we do the real work anyways; the others will just be a nullopt and hence close to zero cost.

Back of the napkin math suggests that even for thousands of entities and hundreds of players we should not be unbearably slow with current CPU speeds, and even if we were, this won't be the limiting factor: Other areas like the O(n²) collision detection will cause performance problems much earlier.

Reply:
I hope you're right that it doesn't matter in practice.

> If each object updates in a step, and each object has, say sqrt(n) children on average (worst case), this gives us O(n sqrt(n) m) for a server step.

But this case practically never happens. My gripe is that you're paying the cost recalculating all the observers *every step*.